metricbeat/module/prometheus/query/data.go

```diff
@@ -0,0 +1,161 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package query
+
+import (
+	"encoding/json"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+// ArrayResponse is for "scalar", "string" type.
+type ArrayResponse struct {
+	Status string    `json:"status"`
+	Data   arrayData `json:"data"`
+}
+type arrayData struct {
+	ResultType string        `json:"resultType"`
+	Results    []interface{} `json:"result"`
+}
+
+// MapResponse is for "vector", "matrix" type from Prometheus Query API Request
+type MapResponse struct {
+	Status string  `json:"status"`
+	Data   mapData `json:"data"`
+}
+type mapData struct {
+	ResultType string      `json:"resultType"`
+	Results    []mapResult `json:"result"`
+}
+type mapResult struct {
+	Metric  map[string]string `json:"metric"`
+	Vector  []interface{}     `json:"value"`
+	Vectors [][]interface{}   `json:"values"`
+}
+
+func parseResponse(body []byte, pathConfig QueryConfig) ([]mb.Event, error) {
+	var events []mb.Event
+	var resultType string
+	var convertedMap MapResponse
+
+	// try to convert to array response
+	convertedArray, err := convertJSONToArrayResponse(body)
+	if err != nil {
+		return events, err
+	}
+	resultType = convertedArray.Data.ResultType
+
+	// check if it is a vector or matrix and unmarshal more
+	if resultType == "vector" || resultType == "matrix" {
+		convertedMap, err = convertJSONToMapResponse(body)
+		if err != nil {
+			return events, err
+		}
+		resultType = convertedMap.Data.ResultType
+	}
```

Comment:
I'll introduce an else here to return "unknown type" or whatever. Just in case

Reply:
It is handled in the `default` case of the `switch`.

Reply:
It's not important anyways, I forgot to write the "nit". So the question is why isn't this pieces of code inside the switches directly?

I feel that all this questions are a consequence of the scope of this function, which is too big and must be split in smaller functions. You can also create a smaller type:

```go
type ResultType struct {
    Data   struct {
        ResultType string  `json:"resultType"`
    } `json:"data"`
}
```

Just to get the type instead of attempting to parse all. Then you create the switch:

```go
switch resultType.Data.ResultType:
case "matrix", "vector":
    ev, err := getMbEventFromMatrix(body)
    if err != nil { errors.Wrapf...}
    events = append(events, ev)
case "scalar", "string":
    ev, err := getMbEventFromScalarOrString(body)
    if err != nil { errors.Wrapf...}
    events = append(events, ev)
```